tsfresh/feature_extraction/extraction.py

```diff
@@ -185,36 +186,39 @@ def _extract_features_parallel_per_sample(kind_to_df_map, settings, column_id, c
     :return: The (maybe imputed) DataFrame containing extracted features.
     :rtype: pandas.DataFrame
     """
+
     partial_extract_features_for_one_time_series = partial(_extract_features_for_one_time_series,
                                                            column_id=column_id,
                                                            column_value=column_value,
                                                            settings=settings)
     pool = Pool(settings.n_processes)
 
-    # Submit map jobs per kind per sample
-    results_fifo = Queue()
-    for kind, df_kind in kind_to_df_map.items():
-        df_grouped_by_id = df_kind.groupby(column_id)
-        results_fifo.put(
-            pool.map_async(
-                partial_extract_features_for_one_time_series,
-                [(kind, df_group) for _, df_group in df_grouped_by_id],
-                chunksize=settings.chunksize
-            )
-        )
+    # Submit map jobs per kind per sample in a generator object
+    kind_and_df_group_list = ((kind, df_group)
+                              for kind, df_kind in kind_to_df_map.items()
+                              for _, df_group in df_kind.groupby(column_id))
+
+    dfs_per_kind = pool.map(
+        partial_extract_features_for_one_time_series,
+        kind_and_df_group_list,
+        chunksize=settings.chunksize,
+    )
 
+    # Wait for the pool to finish its job
     pool.close()
+    pool.join()
 
-    # Wait for the jobs to complete and concatenate the partial results
-    dfs_per_kind = []
-    while not results_fifo.empty():
-        map_result = results_fifo.get()
-        dfs = map_result.get()
-        dfs_per_kind.append(pd.concat(dfs, axis=0).astype(np.float64))
+    # dfs_per_kind is now a list of (kind, df). We want to extract all dfs with the same kind and concatenate them
+    # We have to make sure to not create unnecessary copies here.
+
+    # We will use groupby to group the list by their kind entry. The keyfunction extracts this kind for us.
+    keyfunction = lambda kind_and_df: kind_and_df[0]
```

Comment:
lets refactor `keyfunction` to maybe `first_i` or use anynymous functions `lambda x: x[0]`. keyfunction kind of implies that it is generating keys